src/channel.rs

```diff
@@ -58,11 +58,29 @@ impl<'a> Channel<'a> {
 
 impl fmt::Debug for ChannelData {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("ChannelData")
-            .field("id", &self.id)
-            .field("binary", &self.binary)
-            .field("data", &self.data.len())
-            .finish()
+        let mut ds = f.debug_struct("ChannelData");
+
+        ds.field("id", &self.id);
+        ds.field("binary", &self.binary);
+
+        if self.binary {
+            ds.field("data", &self.data.len());
+        } else {
+            match str::from_utf8(&self.data) {
+                Ok(s) => {
+                    if s.is_ascii() {
+                        ds.field("data", &s);
+                    } else {
+                        ds.field("data", &self.data.len());
+                    }
+                }
+                Err(e) => {
+                    ds.field("data", &e);
+                }
+            }
+        }
+
+        ds.finish()
```

Comment:
Well if you log something call `ChannelData` it doesn't seem unreasonable for it to include the data so that seems okay. It might be very noisy for the logs since these messages can contain quite a lot of data though.

Do we currently rely on this somewhere in str0m?

Reply:
Maybe we can substring it? First 30 chars or something?

Reply:
I actually initially did the first 10 (first 7 + ... if longer) but it was a bit fiddly and not clear to me that `Debug` shouldn't just dump everything the struct's got.

Reply:
In some respects we already accepted that it doesn't dump everything, since we do `.len()` instead of the actual data. I think if you iterate over the chars and `take(10)`, you can quite easily output them one by one.

Reply:
We can always add a bespoke `dbg_fmt_full` method as an alternative

Reply:
if it's not logged within str0m for info+ level it should be fine.

Reply:
When I say fiddly, I mean that it didn't seem straightforward at all to render some kind of `Debug` text that indicated that the suffix of the string was elided -- the alternatives are:
* pretend that the data is only 10 characters long (confusing)
* pretend that the data actually contains an ellipsis (confusing)
* mark the struct as not being 'exhaustive' thereby putting a `..` after the field (confusing)
* implement an alternative `String` `Debug` instance just for this case (a lot of work for... a few bytes of less-than-complete output)

At that point, I got sad and decided that anyone using `Debug` should just deal with the flood of data and if they didn't want that they should use `Display` or write their own printer or something.

Reply:
I think an ellipsis is good!

I use debug! quite a bit, and I want the output to be orienting and detailed, but not down to the exact packet/binary/data level (that's trace as I see it).

Reply:
The missing piece was `format_args!`.